### ISSUE TITLE
Comment in glue() range causes parse failure in object_usage_linter

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1919]
```r
lint("
foo <- function() {
  x <- 2
  glue::glue(
    '{x}' # a comment
  )
}
", object_usage_linter())
# Error: Linter 'object_usage_linter' failed in /tmp/RtmpAirXij/file39151e5be6f486: Unexpected failure to parse glue call, please report: <text>:2:0: unexpected end of input
# 1: glue::glue('{x}'# a comment)
#    ^
```

**Comment by MichaelChirico** [comment]
The trouble is here:

https://github.com/r-lib/lintr/blob/c69f7926e2b01d06171dfe63bef6a90fd9a656f8/R/object_usage_linter.R#L191

`xml_text()` will collapse all the text across the nodes, ignoring newlines, hence trying to parse `glue::glue('{x}'#a comment)`.

What we'd ideally do is exclude any `COMMENT` nodes below each `glue_call` before attempting to parse.

@jimhester could you offer some guidance/`xml2` wizardry here? At a glance `xml_remove()` is what we want, but we don't want to permanently modify the parse tree, is there an easy way to do that? An alternative would be to set `xml_text()` of the `COMMENT` nodes to `''`, then re-set it `on.exit()`, which feels a bit clunky too.

Related again to https://github.com/r-lib/xml2/issues/341#issuecomment-1100794847 :)

**Comment by AshesITR** [comment]
> I would use something like this to get just the text in the STR_CONST nodes, rather than any node type.
> 
> ```r
> xml2::xml_text(xml2::xml_find_first(glue_calls, "//STR_CONST/text()"))
> ```

This will break glue calls with multiple fragments, e.g.

```r
glue::glue(
  "This is the first glued line with 1 + 1 = {1 + 1}\n",
  "This is the second line with pi = {pi}"
)
```

**Comment by MichaelChirico** [comment]
and a find_all approach will also have to ignore keyword arguments.

I think it's doable but feels like a workaround /much messier than `glue_calls |> xml_exclude(comments) |> xml_text()`

**Comment by MichaelChirico** [comment]
`xml2lang()` :)

not a bad idea, this thing with comments definitely comes up here & there

**Comment by jimhester** [comment]
```r
paste0(collapse = "", xml2::xml_text(xml2::xml_find_all(glue_calls, "./*[not(self::COMMENT)]")))
```

**Comment by MichaelChirico** [comment]
ah, great minds :)

```r
xml <- parse(text = "glue(\n'{x}' # comment\n)") |>
  xml_parse_data() |>
  read_xml()

xml |>
  xml_find_all(".//*[not(self::COMMENT or self::expr or self::exprlist)]") |>
  xml_text() |>
  paste(collapse = "") |>
  str2lang()
# glue("{x}")
```

`./*` will struggle with more complicated expressions I think:

```r
glue(
  format(
    Sys.Date(),
    "%F {x}" # YYYY-MM-DD, then 'x'
  )
)
```
